### PR TITLE
Add support for GitLab OAuth2 authentication 

### PR DESCRIPTION
**Pull request by fcelda**
This patch set adds a support for OAuth2 authentication with [GitLab](http://doc.gitlab.com/ce/integration/oauth_provider.html).

### CODE REVIEW

**Comment by tardyp** [comment]
cool! :+1:

**Comment by tardyp** [comment]
an entry in the relnote, and I will be very happy

**Comment by fcelda** [comment]
I've added the entry. But I didn't manage to make the existing tests for OAuth work. Thus the low coverage.

**Comment by tardyp** [comment]
well, you could easily unit test getUserInfoFromOAuthClient(), at least run the code, and make sure it does not crash, appart from that this is more an integration test which would be needed, and this is much more difficult to build

**Comment by fcelda** [comment]
OK then. I'll check what can be done. :)

**Comment by fcelda** [comment]
I've added both unit tests and E2E tests. As for the second ones, is't kind of a blind change. I couldn't make the existing E2E tests work. The code passes the authentication class constructor. But then it just hangs.

**Comment by tardyp** [comment]
Oh, I did not remember I had e2e tests for oauth2. great!

**Comment by tardyp** [comment]
> But then it just hangs.
Ah. I did not read until the end.. As those tests are not run continuously, they are probably broken :-(

File: master/docs/relnotes/index.rst [hunk 1]
```diff
@@ -21,6 +21,7 @@ Features
 ~~~~~~~~
 
 * :class:`GitPoller` now has a ``buildPushesWithNoCommits`` option to allow the rebuild of already known commits on new branches.
+* Add GitLab authentication plugin for web UI. See :class:`buildbot.www.oauth2.GitLabAuth`.
```
Comment (rutsky) [inline master/docs/relnotes/index.rst]: @tardyp I believe this should go into 0.9.0b8.rst release notes.

I created template (and filled it with my stuff) in PR #2012.

Comment (tardyp) [inline master/docs/relnotes/index.rst]: No. The next release note is created at the time of release.

Le mar. 23 févr. 2016 03:36, Vladimir Rutsky <notifications@github.com> a
écrit :

Comment (rutsky) [inline master/docs/relnotes/index.rst]: Oh, fixed in PR #2012, thanks!

**Comment by asedeno** [comment]
FYI, I suspect this is affected by http://trac.buildbot.net/ticket/3439, which would cause the username to clobber the email address in the userinfo dict.

**Comment by fcelda** [comment]
@asedeno I noticed that problem on 0.9.0b7 but had no time to verify on master.

**Comment by rodrigc** [comment]
@tardyp @fcelda I am trying to test OAuth2 in buildbot in order to solve https://github.com/buildbot/buildbot/issues/3374 filed by @zware .  How do I test OAuth2?  I tried creating a file named /tmp/a.json as specified in https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/unit/test_www_oauth.py#L453 with my Google login/pw,

then I tried running:

```
env OAUTHCONF=/tmp/a.json trial  buildbot.test.unit.test_www_oauth.OAuth2AuthGoogleE2E
```

but the test failed for me.

Any ideas as to what I am doing wrong?

Thanks.

**Pull request merged**